c10/core/QEngine.h

```diff
@@ -15,11 +15,13 @@ enum class QEngine : uint8_t {
   NoQEngine = 0,
   FBGEMM = 1,
   QNNPACK = 2,
+  QMKLDNN = 3,
```

Comment:
I don't think this as a separate enum makes sense because MKLDNN and FBGEMM are not exclusive.

First of all, this engine mechanism was added just for testing or debugging purposes - in general our default path should route to the fastest implementation that is available on the current platform.

And for test/debugging purposes - we probably need a more fine-grained mechanism. For example, set of "allowed engines" instead of a single value. Let's have more discussion on it.

Reply:
> First of all, this engine mechanism was added just for testing or debugging purposes - in general our default path should route to the fastest implementation that is available on the current platform.

@dzhulgakov Thanks for the suggestion. To get the performance benefit of what current MKL-DNN supports, i.e. AQ with zero point 0 only, do you mind we prepack both MKL-DNN and FBGEMM weights so that we can have a way to pick the faster path at runtime and fall back to FBGEMM where MKL-DNN does not support? This would provide us flexibility for getting the best performance even after MKL-DNN supports complete AQ - consider MKL-DNN is faster on some input shapes while FBGEMM is faster on others. The downside is that there would be more memory needed to hold additional prepacked weights. We can share more performance numbers if you think this direction is OK.